packages/@atjson/document/src/document.ts

```diff
@@ -684,6 +691,20 @@ export class Document {
     );
   }
 
+  private withStableIds(seed: string) {
+    let ids: Record<string, string> = {};
+    let counter = 1;
+    for (let annotation of this.annotations) {
+      ids[annotation.id] = `${seed}-${counter++}`;
```

Comment:
doesn't this make it so the annotation -> stable id mapping depends on the order of the annotations in the array?

Reply:
Yeah, I think you should make a test that if you take a document and reorder its annotations without changing anything else, it still compares equal

Reply:
sounds good!

Reply:
+1 this – I'm not sure I've got my head fully wrapped around all this, but would it make sense for this to be stable even if the set of annotations varies across documents? The most overkill approach would be to hash the annotation, but it might actually be faster to do that than come up with some more bespoke serialization?

Reply:
@blaine the problem is that, with references, there's no way to tell if two "equivalent" annotations are isomorphic under document-wide reference-renaming. That is to say, we would need some kind of clever graph traversal to even generate the appropriate hash value, which is the thing we're trying to avoid lol

Reply:
the solution here is to figure out a way to rename all the annotation IDs in a way that will work out the same for any documents that pass the equality check, by simply ordering the annotations in some deterministic way and numbering them sequentially

Reply:
@blaine I'm down with a hash approach! I'm not sure how circular references would work with a hashing approach

Reply:
@colin-alexa thanks for the very nice explanation :)

Reply:
@colin-alexa this means that this scheme depends on the ordering, no? In particular, if an ordering is not stable over all initial re-orderings then you will have issues.

The comparator function we look like we are using is this function: https://github.com/CondeNast/atjson/blob/latest/packages/%40atjson/document/src/collection.ts#L10 which does not satisfy that condition

Reply:
that comparator only considers `start` `end` `type` and `rank`. Perhaps that's good enough. But that means you have an indeterminate ordering for two annotations of the same type at the same position with possibly different attributes

Reply:
that's unfortunate, but is an existing problem with sorting annotations, so I think this is not making things worse...